src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -0,0 +1,360 @@
+                        serverCertificate,
+                        chain,
+                        sslPolicyErrors);
+                    if (!result)
+                    {
+                        throw WinHttpException.CreateExceptionUsingError(
+                            (int)Interop.WinHttp.ERROR_WINHTTP_SECURE_FAILURE);
+                    }
+                }
+                finally
+                {
+                    if (chain != null)
+                    {
+                        chain.Dispose();
+                    }
+                }
+            }
+        }
+
+        private static void OnRequestError(WinHttpRequestState state, Interop.WinHttp.WINHTTP_ASYNC_RESULT asyncResult)
+        {
+            WinHttpTraceHelper.TraceAsyncError("OnRequestError", asyncResult);
+            
+            Debug.Assert(state != null, "OnRequestError: state is null");
+
+            var innerException = WinHttpException.CreateExceptionUsingError((int)asyncResult.dwError);
+
+            switch ((uint)asyncResult.dwResult.ToInt32())
+            {
+                case Interop.WinHttp.API_SEND_REQUEST:
+                    state.TcsSendRequest.TrySetException(innerException);
+                    break;
+                    
+                case Interop.WinHttp.API_RECEIVE_RESPONSE:
+                    if (asyncResult.dwError == Interop.WinHttp.ERROR_WINHTTP_RESEND_REQUEST)
+                    {
+                        state.RetryRequest = true;
+                        state.TcsReceiveResponseHeaders.TrySetResult(false);
+                    }
+                    else if (asyncResult.dwError == Interop.WinHttp.ERROR_WINHTTP_CLIENT_AUTH_CERT_NEEDED)
+                    {
+                        // WinHttp will automatically drop any client SSL certificates that we
+                        // have pre-set into the request handle.  For security reasons, we don't
+                        // allow the certificate to be re-applied. But we need to tell WinHttp
+                        // that we don't have any certificate.
+                        WinHttpHandler.SetNoClientCertificate(state.RequestHandle);
+                        state.RetryRequest = true;
+                        state.TcsReceiveResponseHeaders.TrySetResult(false);
+                    }
+                    else if (asyncResult.dwError == Interop.WinHttp.ERROR_WINHTTP_OPERATION_CANCELLED)
+                    {
+                        state.TcsReceiveResponseHeaders.TrySetCanceled(state.CancellationToken);
+                    }
+                    else
+                    {
+                        state.TcsReceiveResponseHeaders.TrySetException(innerException);
+                    }
+                    break;
+
+                case Interop.WinHttp.API_READ_DATA:
+                    if (asyncResult.dwError == Interop.WinHttp.ERROR_WINHTTP_OPERATION_CANCELLED)
+                    {
+                        // TODO: Which cancellation token to pass in?
```

Comment:
The cancellation token itself is something pass in by the caller (3rd party dev).  Different tokens could be passed in for the handler SendAsync() vs. the response stream ReadAsync() methods.

I'll revise the TODO.  But basically it is for me to figure out how to plumb in the caller's cancellation token from ReadAsync() and get cancellation honored.